src/crypto/openssl/openssl_crypto_accel.cc

```diff
@@ -74,7 +76,11 @@ bool evp_transform(unsigned char* out, const unsigned char* in, size_t size,
   ceph_assert(len_final == 0);
   return (len_update + len_final) == static_cast<int>(size);
 }
-                        
+
+OpenSSLCryptoAccel::OpenSSLCryptoAccel()
+    : engine_id(g_ceph_context->_conf->openssl_accel_engine),
+      engine(OpensslAccelEngine::get_instance(engine_id).get_engine()) {}
```

Comment:
@cbodley @rosinL 

can we enable `plugin_crypto_accelerator` to pass additional parameters to the plugin. for instance, currently we have 3 crypto plugins: "openssl", "isa-l", and "qat". in which, openssl supports multiple engines for encryption and digesting. the openssl offers both builtin engines and a machinery for loading dynamic engines plugins. and the runtime behavior of engines can be customized using `ENGINE_ctrl*` functions. so i am wondering if we can use a delimiter (":" for example) in `plugin_crypto_accelerator` to separate the name of the plugin and the parameters to be passed to the plugin?

in this case, if we have a setting like
```
[client.rgw]
plugin_crypto_accelerator = openssl:openssl:cmd1=arg1,cmd2=arg2
```
openssl plugin will add and use its builtin openssl engine. and initialize this engine by calling
```c++
ENGINE_ctrl_cmd_string(engine, "cmd1", "args1", nullptr);
ENGINE_ctrl_cmd_string(engine, "cmd2", "args2", nullptr);
```

i believe the default engine currently being used by openssl crypto plugin is `ENGINE_get_cipher_engine(NID_aes_256_cbc)`

so we can use "aes_256_cbc" as the default openssl engine, if it's not specified.

what do you think?

Reply:
@tchaikov ,  sorry for late reply.  it seems that there is an easy way to configure stuff like engine (with all its cmd string) using openssl configuration files.  
all we need do is calling `OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CONFIG, NULL)` once for one process(like RGW), then openssl will try to load config through local environment variable 
`OPENSSL_CONF` (if not set, using its default conf maybe) itself, which point to one customized file. and one typical way to configure engine is like: (from [here](https://www.openssl.org/docs/man1.1.1/man5/config.html))
```
 [engine_section]

 # Configure ENGINE named "foo"
 foo = foo_section

 [foo_section]
 # Load engine from DSO
 dynamic_path = /some/path/fooengine.so
 # A foo specific ctrl.
 some_ctrl = some_value
 # Another ctrl that doesn't take a value.
 other_ctrl = EMPTY
 # Supply all default algorithms
 default_algorithms = ALL
```
we can even control the engine only work for specific algorithm (like engine1 for MD5 only).

what's your opinion about this? if you think it is not bad, is there still a need to bring these configures to ceph, for some reason like cluster management?

Reply:
@egggHang my major concern is ease of management. what's your plan to populate `openssl.cnf` to multiple OSD hosts?

Reply:
@tchaikov  maybe we can add a new config to ceph , `“openssl_engine_options”` for example,  then pass all parameters and seperate paras for one engine by `,` and `:` between engines, like:

`openssl_engine_options = "engine=engine1,dynamic_path=/some/path/engine1.so,default_algorithms=MD5:engine=engine2,dynamic_path=/some/path/engine2.so,default_algorithms=SHA1,other_ctrl=other_value"`

then we add a new function to parse this, maybe faking a openssl-style conf with several section and calling `CONF_modules_load` or `int_engine_configure` in openssl to finish leftover task, or just using basic `ENGINE_by_id` or `ENGINE_ctrl_cmd_string` to parse in a loop.  I prefer the former one. besides, we need to make sure all this run once, like what openssl does now.

any advice about this?